lm_eval/models/huggingface.py

```diff
@@ -678,39 +678,61 @@ def _create_tokenizer(
             )
         return None
 
-    def _detect_batch_size(self, requests=None, pos: int = 0):
-        if requests:
+    def _detect_batch_size(self, requests=None, pos: int = 0) -> int:
+        if len(requests[0]) == 3:  # logprob evals
             _, context_enc, continuation_enc = requests[pos]
             max_length = len(
                 (context_enc + continuation_enc)[-(self.max_length + 1) :][:-1]
             )
             max_context_enc = len(context_enc[-(self.max_length + 1) :])
             max_cont_enc = len(continuation_enc[-(self.max_length + 1) :])
-        else:
-            max_length = self.max_length
+            security_margin_factor = (
+                4  # batch sizes for log prob evals sometimes generate OOMs
+            )
+        elif len(requests[0]) == 2:  # generative evals
+            # using rolling window with maximum context
```

Comment:
As a suggestion of PR improvement. Please add more thorough comment about why max_length detection is not universal across task types (logptob vs generative).